### PR TITLE
[no-magic-numbers] option "allow-element-access" and ignore nums w/ comments

### PR DESCRIPTION
**Pull request by aervin**
#### PR checklist

- [x] Addresses an existing issue: [#2721](https://github.com/palantir/tslint/issues/2721) [#1883](https://github.com/palantir/tslint/issues/1883)
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [x] Documentation update

#### Overview of change:
* By default, the rule ignores "magic" numbers if they are on the same line as a comment.
* Adds support for config object
* Adds new config option `allow-element-access`
* Moves `allowed-numbers` config option into config object

#### Is there anything you'd like reviewers to focus on?
The way this rule gathers the source file's comments--is it sufficient? Are there edge cases I'm forgetting? See: the `generateCommentPositions` method.

### CODE REVIEW

File: src/rules/noMagicNumbersRule.ts [hunk 1]
```diff
@@ -31,15 +41,32 @@ export class Rule extends Lint.Rules.AbstractRule {
...
+                    minLength: 1,
+                },
+                "allow-element-access": {
+                    type: "boolean",
+                },
             },
-            minLength: 1,
         },
-        optionExamples: [true, [true, 1, 2, 3]],
+        optionExamples: [true, [true, 1, 2, 3], [true, {
```
Comment (giladgray) [inline src/rules/noMagicNumbersRule.ts]: this second example no longer matches the schema, which makes this PR a **breaking change**

File: src/rules/noMagicNumbersRule.ts [hunk 2]
```diff
@@ -62,15 +89,55 @@ export class Rule extends Lint.Rules.AbstractRule {
 
     public static DEFAULT_ALLOWED = [ -1, 0, 1 ];
 
+    /* tslint:disable no-unsafe-any */
```
Comment (giladgray) [inline src/rules/noMagicNumbersRule.ts]: wait why... can we not disable this rule here?

File: src/rules/noMagicNumbersRule.ts [hunk 3]
```diff
@@ -62,15 +89,55 @@ export class Rule extends Lint.Rules.AbstractRule {
...
-        return this.applyWithWalker(new NoMagicNumbersWalker(sourceFile, this.ruleName, new Set(allowedNumbers.map(String))));
+        return this.applyWithWalker(
+            new NoMagicNumbersWalker(
+                sourceFile, this.ruleName,
+                this.parseOptions(this.ruleArguments[0]),
+            ),
+        );
+    }
+
+    private parseOptions(args: any): Options {
```
Comment (giladgray) [inline src/rules/noMagicNumbersRule.ts]: from code, looks like this is `args?: number[] | { [key: string]: boolean }`

File: src/rules/noMagicNumbersRule.ts [hunk 4]
```diff
@@ -17,10 +17,20 @@
...
+    forEachComment,
+    isCallExpression,
+    isElementAccessExpression,
+    isIdentifier,
+    isSameLine } from "tsutils";
 import * as Lint from "../index";
 import { isNegativeNumberLiteral } from "../language/utils";
 
+interface Options {
+    allowedNumbers: number[] | Set<string>;
```
Comment (giladgray) [inline src/rules/noMagicNumbersRule.ts]: `Set<string>` does not make sense for numbers. why is that here and can it be just `number[]` (or `Set<number>`)?

File: src/rules/noMagicNumbersRule.ts [hunk 5]
```diff
@@ -62,15 +89,55 @@ export class Rule extends Lint.Rules.AbstractRule {
...
     }
+    /* tslint:enable no-unsafe-any */
 }
 
-class NoMagicNumbersWalker extends Lint.AbstractWalker<Set<string>> {
+class NoMagicNumbersWalker extends Lint.AbstractWalker<Options> {
     public walk(sourceFile: ts.SourceFile) {
         const cb = (node: ts.Node): void => {
+            if (
+                commentPositions(this.sourceFile)
```
Comment (giladgray) [inline src/rules/noMagicNumbersRule.ts]: ❌ strongly against this feature. instead of implicitly disabling the rule if a comment appears, users should explicitly use the **existing `// tslint:disable-next-line` feature**, which is always a comment itself!

File: src/rules/noMagicNumbersRule.ts [hunk 6]
```diff
@@ -87,8 +154,35 @@ class NoMagicNumbersWalker extends Lint.AbstractWalker<Set<string>> {
...
             this.addFailureAtNode(node, Rule.FAILURE_STRING);
         }
     }
 }
+
+/**
+ * Used for checking when a magic number is on the same line as
+ * a comment. In such a case, the magic number should be ignored.
+ */
+function commentPositions(sourceFile: ts.SourceFile): number[] {
```
Comment (giladgray) [inline src/rules/noMagicNumbersRule.ts]: ❌ kill this feature please.

File: src/rules/noMagicNumbersRule.ts [hunk 7]
```diff
@@ -87,8 +154,35 @@ class NoMagicNumbersWalker extends Lint.AbstractWalker<Set<string>> {
     }
 
     private checkNumericLiteral(node: ts.Node, num: string) {
-        if (!Rule.ALLOWED_NODES.has(node.parent!.kind) && !this.options.has(num)) {
+        if (
+            /* "allow-element-access" logic */
+            !(
+                node.parent !== undefined &&
+                isElementAccessExpression(node.parent) &&
+                this.options.allowElementAccess
```
Comment (giladgray) [inline src/rules/noMagicNumbersRule.ts]: please pull this `()` out to a local variable so it's easier to follow this logic.

File: test/rules/no-magic-numbers/default/test.ts.lint [hunk 8]
```diff
@@ -28,3 +34,18 @@ class A {
 enum Test {
     key = 1337,
 }
+
+it('should provide a link to the landing page', () => {
```
Comment (giladgray) [inline test/rules/no-magic-numbers/default/test.ts.lint]: while it's cool to put real code here, it's also very confusing when the only thing we care about is that little index operator inside the _next_ line.

please remove the `it()` wrapper and just test the array access.

File: src/rules/noMagicNumbersRule.ts (hunk 1 above)
Comment (giladgray) [reply src/rules/noMagicNumbersRule.ts]: though i see you've preserved support for the old usage below.

**Comment by giladgray** [comment]
ok cool will close. thanks for the info @aervin!

fwiw, you can totally add a comment after the disable:
```
// tslint:disable-next-line:no-magic-numbers the number is magic because Snape said so
```

File: src/rules/noMagicNumbersRule.ts (hunk 5 above)
Comment (aervin) [reply src/rules/noMagicNumbersRule.ts]: I see a big difference here:

```ts
foo = bar[1]; // some helpful info about "1"
```

and
```ts
// tslint:disable-next-line no-magic-numbers
foo = bar[1];
```

Please keep in mind that this PR is in response to issues about the strictness of no-magic-numbers.

**Comment by aervin** [comment]
Alright thanks @giladgray
